service/raft/group0_state_id_handler.cc

```diff
@@ -6,15 +6,124 @@
  * SPDX-License-Identifier: AGPL-3.0-or-later
  */
 #include "service/raft/group0_state_id_handler.hh"
+#include "db/config.hh"
+#include "service/raft/raft_address_map.hh"
+#include "compaction/compaction_manager.hh"
 #include "gms/gossiper.hh"
 #include "log.hh"
+#include "raft/server.hh"
+#include "replica/database.hh"
+#include "service/raft/raft_group_registry.hh"
 
 namespace service {
 
 static logging::logger slogger("group0_tombstone_gc_handler");
 
-group0_state_id_handler::group0_state_id_handler(gms::gossiper& gossiper)
-    : _gossiper(gossiper) {
+raft::server* group0_server_accessor::get_server() const {
+    return _raft_gr.find_server(_group0_id);
+}
+
+lowres_clock::duration group0_state_id_handler::get_broadcast_interval(const replica::database& db) {
+    return std::chrono::milliseconds{db.get_config().group0_tombstone_gc_interval_in_ms()};
+}
+
+group0_state_id_handler::group0_state_id_handler(
+        replica::database& local_db, gms::gossiper& gossiper, const raft_address_map& address_map, group0_server_accessor server_accessor)
+    : _local_db(local_db)
+    , _gossiper(gossiper)
+    , _address_map(address_map)
+    , _server_accessor(server_accessor)
+    , _check_interval(get_broadcast_interval(local_db)) {
+}
+
+group0_state_id_handler::~group0_state_id_handler() {
+    if (!_stopped) {
+        on_fatal_internal_error(slogger, "group0_state_id_handler is being destroyed without being stopped");
+    }
+}
+
+void group0_state_id_handler::start() {
+    if (this_shard_id() != 0) {
+        on_fatal_internal_error(slogger, "group0_state_id_handler must be started on shard 0");
+    }
+
+    if (_stopped) {
+        return;
+    }
+
+    _timer.set_callback([this] {
+        std::invoke([this] {
+            auto* const group0_server = _server_accessor.get_server();
+            if (!group0_server) {
+                slogger.debug("Group0 server not found");
+                return;
+            }
+
+            const auto& group0_members = group0_server->get_configuration().current;
```

Comment:
We can just skip the round (or retry in a few seconds/minutes) in case we are in the middle of a config change.

Reply:
Done, just joining the configs (we could also skip based on `is_joint()` I guess, but it is likely that it might be skipped for some other reason anyway - e.g. missing the endpoint or state_id information ... and if the info is all there, we might even succeed)